extensions/rvadapter/src/main/java/com/google/android/agera/rvadapter/RepositoryAdapter.java

```diff
@@ -318,13 +337,51 @@ public final void startObserving() {
    */
   public final void stopObserving() {
     observable.removeUpdatable(this);
+    updateAllObservable.removeUpdatable(updateAllUpdatable);
   }
 
   /**
-   * Invalidates the data set so {@link RecyclerView} will schedule a rebind of all data.
+   * Invalidates the data set so {@link RecyclerView} will schedule rebinds of changed data.
    */
   @Override
   public final void update() {
+    if (dataInvalid) {
+      return;
+    }
+
+    int totalItemsAddedCount = 0;
+    for (int i = 0; i < repositoryCount; i++) {
+      endPositions[i] += totalItemsAddedCount;
+
+      final Object newData = repositories[i].get();
+      if (!newData.equals(data[i])) {
+        data[i] = newData;
+        final int itemCount = presenters[i].getItemCount(data[i]);
+        final int startPosition = i > 0 ? endPositions[i-1] : 0;
+        final int endPosition = startPosition + itemCount;
+        final int itemsAddedCount = endPosition - endPositions[i];
+        if (itemsAddedCount == 0) {
+          notifyItemRangeChanged(startPosition, itemCount);
+        } else {
+          totalItemsAddedCount += itemsAddedCount;
+          endPositions[i] += itemsAddedCount;
+
+          if (itemsAddedCount > 0) {
+            notifyItemRangeChanged(startPosition, itemCount - itemsAddedCount);
```

Comment:
I've made this behavior opt-in for the cases where `!newData.equals(data[i])` isn't valid. `RepositoryPresenterCompiler` could probably be extended with an `areItemsDifferent(Merger)` similar to `.notifyIf()`.

`itemRangeChanged` + `itemRangeInserted/itemRangeRemoved` is adequate for all cases but perhaps not optimal. Allowing clients to tweak this isn't straightforward and `RepositoryAdapter` would need to be more stateful which I'm not too keen on. Although I was considering this from looking at the whole data set together, but I'm now thinking maybe we can do something in the `RepositoryPresenter` as the different 'lists' update independently of each other.